src/librustc_mir/dataflow/impls/use_def_chain.rs

```diff
@@ -0,0 +1,302 @@
+/// for every use in the program. This allows consumers to traverse the graph of data dependencies
+/// efficiently. For example, a `UseDefChain` could be used for const-propagation by iterating over
+/// the definitions which could possibly reach `y` at the exit of the following function and
+/// realizing that all of them are the constant `6`.
+///
+/// ```rust,no_run
+/// fn six(condition: bool) -> i32 {
+///     let mut y;
+///     let a = 2 * 3;
+///     let b = 10 - 4;
+///
+///     if condition {
+///         y = a;
+///     } else {
+///         y = b;
+///     }
+///
+///     y
+/// }
+/// ```
+pub struct UseDefChain<'me, 'tcx> {
+    reaching_defs: &'me DataflowResults<'tcx, ReachingDefinitions>,
+
+    /// Maps each local that is used at a given location to the set of definitions which can reach
+    /// it.
+    defs_for_use: DefsForUseAtLocation,
+}
+
+impl<'me, 'tcx> UseDefChain<'me, 'tcx> {
+    pub fn new(
+        body: &'me mir::Body<'tcx>,
+        reaching_defs: &'me DataflowResults<'tcx, ReachingDefinitions>,
+        observed_addresses: &DataflowResults<'tcx, HaveBeenBorrowedLocals<'_, 'tcx>>,
+    ) -> Self {
+        let defs = reaching_defs.operator();
+        let all_defs_for_local = all_defs_for_local(body, defs, &observed_addresses);
+        let locals_used_in_block = LocalsUsedInBlock::new(body);
+        let curr_defs_for_local =
+            IndexVec::from_elem(BitSet::new_empty(defs.len()), &body.local_decls);
+
+        let mut defs_for_use = DefsForUseAtLocation::default();
+
+        {
+            let mut builder = UseDefBuilder {
+                body,
+                ud_chain: &mut defs_for_use,
+                reaching_defs,
+                locals_used_in_block,
+                curr_defs_for_local,
+                all_defs_for_local,
+            };
+
+            builder.visit_body(body);
+        }
+
+        UseDefChain {
+            defs_for_use,
+            reaching_defs,
+        }
+    }
+
+    pub fn defs_for_use(
+        &self,
+        local: Local,
+        location: Location,
+    ) -> impl Iterator<Item = &'_ Definition> + Captures<'tcx> + '_ {
+        let UseDefChain { defs_for_use, reaching_defs, .. } = self;
+
+        defs_for_use
+            .get(&location)
+            .and_then(|for_local| for_local.get(&local))
+            .into_iter()
+            .flat_map(|defs| defs.iter())
+            .map(move |def| reaching_defs.operator().get(def))
+    }
+}
+
+struct UseDefBuilder<'a, 'me, 'tcx> {
+    ud_chain: &'a mut DefsForUseAtLocation,
+
+    body: &'me mir::Body<'tcx>,
+
+    reaching_defs: &'me DataflowResults<'tcx, ReachingDefinitions>,
+
+    /// It's expensive to clone the entry set at the start of each block for every local and apply
+    /// the appropriate mask. However, we only need to keep `curr_defs_for_local` up-to-date for
+    /// the `Local`s which are actually used in a given `BasicBlock`.
+    locals_used_in_block: LocalsUsedInBlock,
```

Comment:
This is a holdover from a previous version of this PR when `all_defs_for_local` was not cached. It may be faster not to iterate over the `mir::Body` to initialize `locals_used_in_block`.